### PR TITLE
allowed input outside of input region

### PR DESCRIPTION
**Pull request by wmww**
removes all the
```
[2018-06-13 09:09:59.379419] <WARNING> mirserver: surface->transform_point() returned nullopt for motion action
```
warnings, and corrects some behavior (such as touching a surface and dragging outside it)

### CODE REVIEW

File: src/server/frontend_wayland/wl_touch.cpp [hunk 1]
```diff
@@ -79,29 +71,21 @@ void mf::WlTouch::handle_event(MirTouchEvent const* touch_ev, WlSurface* surface
         case mir_touch_action_change:
         {
             auto const transformed = surface->transform_point(point);
-            if (transformed)
+            if (focused_surface_for_ids[touch_id] == transformed.surface)
```
Comment (RAOF) [inline src/server/frontend_wayland/wl_touch.cpp]: This seems wrong (although the existing code *also* seems wrong!). This is going to ensure that a (sub)surface never receives a touch event outside its bounds?

I believe the correct behaviour is for the surface that received the `touch_down` to receive all further events for that touch until `touch_up`. I don't know how that interacts with the input region, however...

File: src/server/frontend_wayland/wl_pointer.cpp [hunk 2]
```diff
@@ -112,15 +112,8 @@ void mf::WlPointer::handle_event(MirPointerEvent const* event, WlSurface* surfac
...
-            if (transformed)
-            {
-                handle_enter(transformed.value().first, transformed.value().second);
-                handle_frame();
-            }
-            else
-            {
-                log_warning("surface->transform_point() returned nullopt for enter action");
-            }
+            handle_enter(transformed.position, transformed.surface);
```
[the remark this answers is not in the document]
Comment (wmww) [reply src/server/frontend_wayland/wl_pointer.cpp]: In general, this function should only be called with input inside the input region, as Mir is already using that to determine which surface to dispatch input to. If there is some case where it's getting input outside the region, clients should be able to handle it.

File: src/server/frontend_wayland/wl_touch.cpp (hunk 1 above)
Comment (wmww) [reply src/server/frontend_wayland/wl_touch.cpp]: I believe you are correct. Let me see what I can do about this.

Comment (wmww) [reply src/server/frontend_wayland/wl_touch.cpp]: Lemme write some tests, so the code I end up with actually works.

File: src/server/frontend_wayland/wl_surface.cpp [hunk 3]
```diff
@@ -104,23 +104,23 @@ bool mf::WlSurface::synchronized() const
...
     // loop backwards so the first subsurface we find that accepts the input is the topmost one
     for (auto child_it = children.rbegin(); child_it != children.rend(); ++child_it)
     {
         auto result = (*child_it)->transform_point(point);
-        if (result)
+        if (result.is_in_input_region)
             return result;
     }
     geom::Rectangle surface_rect = {geom::Point{}, buffer_size_.value_or(geom::Size{})};
     for (auto& rect : input_shape.value_or(std::vector<geom::Rectangle>{{{}, buffer_size()}}))
```
Comment (RAOF) [inline src/server/frontend_wayland/wl_surface.cpp]: Hrm. Now that I look again, this seems wrong?

From [the protocol](https://gitlab.freedesktop.org/wayland/wayland/blob/master/protocol/wayland.xml#L1479): if the input region is unset, it is treated as infinite.

This is relevant for input events which will be delivered even though they're outside the normal bounds of the surface - eg: pointer drag, touch.

File: src/server/frontend_wayland/wl_pointer.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (RAOF) [reply src/server/frontend_wayland/wl_pointer.cpp]: Is `input_shape` communicated up to the rest of Mir? If not, this could receive an event that's inside the surface but not inside the input region and incorrectly send an enter event for it?

[the remark this answers is not in the document]
Comment (AlanGriffiths) [reply src/server/frontend_wayland/wl_pointer.cpp]: The input shape is communicated to the Mir scene, and seems to be working "well enough"[1]. All the Wayland code needs to do is identify the correct subsurface to deliver the event to and transform the co-ordinates accordingly.

[1] I don't remember seeing any tests ensuring correct working, so there are likely some corner cases that need fixing (like this).

File: src/server/frontend_wayland/wl_surface.cpp (hunk 3 above)
Comment (wmww) [reply src/server/frontend_wayland/wl_surface.cpp]: What is the problem? If the input region is unset, for any surface/subsurface, the for loop a few lines down loops over one rect the side of the surface:
```
for (auto& rect : input_shape.value_or(std::vector<geom::Rectangle>{{{}, buffer_size()}}))
```
It will then return a `WlSurface::Positoon` with `is_in_input_region` true. Or am I misunderstanding the issue?

**Comment by wmww** [comment]
All the tests (including newly written ones [here](https://github.com/MirServer/wlcs/pull/53)) seem to be passing. If there are cases you think might cause an issue, describe them and I'll write tests for them and see (probably when I get back next week).
